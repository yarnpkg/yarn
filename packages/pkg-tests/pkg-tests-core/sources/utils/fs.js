/* @flow */

import type {Gzip} from 'zlib';

const fs = require('fs-extra');
const klaw = require('klaw');
const path = require('path');
const tarFs = require('tar-fs');
const tmp = require('tmp');
const zlib = require('zlib');

const miscUtils = require('./misc');

exports.walk = function walk(
  source: string,
  {filter, relative = false}: {|filter?: Array<string>, relative?: boolean|} = {},
): Promise<Array<string>> {
  return new Promise((resolve, reject) => {
    const paths = [];

    const walker = klaw(source, {
      filter: itemPath => {
        if (!filter) {
          return true;
        }

        const stat = fs.statSync(itemPath);

        if (stat.isDirectory()) {
          return true;
        }

        const relativePath = path.relative(source, itemPath);

        if (miscUtils.filePatternMatch(relativePath, filter)) {
          return true;
        }

        return false;
      },
    });

    walker.on('data', ({path: itemPath}) => {
      const relativePath = path.relative(source, itemPath);

      if (!filter || miscUtils.filePatternMatch(relativePath, filter)) {
        paths.push(relative ? relativePath : itemPath);
      }

      // This item has been accepted only because it's a directory; it doesn't match the filter
      return;
    });

    walker.on('end', () => {
      resolve(paths);
    });
  });
};

exports.packToStream = function packToStream(
  source: string,
  // $FlowFixMe Flow doesn't support null default parameters in exact types
  {virtualPath = null}: {|virtualPath?: ?string|} = {},
): Gzip {
  if (virtualPath) {
    if (!path.isAbsolute(virtualPath)) {
      throw new Error('The virtual path has to be an absolute path');
    } else {
      virtualPath = path.resolve(virtualPath);
    }
  }

  const zipperStream = zlib.createGzip();

  const packStream = tarFs.pack(source, {
    map: header => {
      if (true) {
        header.name = path.resolve('/', header.name);
        header.name = path.relative('/', header.name);
      }

      if (virtualPath) {
        header.name = path.resolve('/', virtualPath, header.name);
        header.name = path.relative('/', header.name);
      }

      return header;
    },
  });

  packStream.pipe(zipperStream);

  packStream.on('error', error => {
    zipperStream.emit('error', error);
  });

  return zipperStream;
};

exports.packToFile = function packToFile(target: string, source: string, options: *): Promise<void> {
  const tarballStream = fs.createWriteStream(target);

  const packStream = exports.packToStream(source, options);
  packStream.pipe(tarballStream);

  return new Promise((resolve, reject) => {
    tarballStream.on('error', error => {
      reject(error);
    });

    packStream.on('error', error => {
      reject(error);
    });

    tarballStream.on('close', () => {
      resolve();
    });
  });
};

exports.createTemporaryFolder = function createTemporaryFolder(): Promise<string> {
  return new Promise((resolve, reject) => {
    tmp.dir({unsafeCleanup: true}, (error, dirPath) => {
      if (error) {
        reject(error);
      } else {
        resolve(dirPath);
      }
    });
  });
};

exports.createTemporaryFile = async function createTemporaryFile(filePath: string): Promise<string> {
  if (filePath) {
    if (path.normalize(filePath).match(/^(\.\.)?\//)) {
      throw new Error('A temporary file path must be a forward path');
    }

    const folderPath = await exports.createTemporaryFolder();
    return path.resolve(folderPath, filePath);
  } else {
    return new Promise((resolve, reject) => {
      tmp.file({discardDescriptor: true}, (error, filePath) => {
        if (error) {
          reject(error);
        } else {
          resolve(filePath);
        }
      });
    });
  }
};

exports.writeFile = async function writeFile(target: string, body: string | Buffer): Promise<void> {
  await fs.mkdirp(path.dirname(target));
  await fs.writeFile(target, body);
};

exports.readFile = function readFile(source: string, encoding: ?string = null): Promise<any> {
  return fs.readFile(source, encoding);
};

exports.writeJson = function writeJson(target: string, object: any): Promise<void> {
  return exports.writeFile(target, JSON.stringify(object));
};

exports.readJson = async function readJson(source: string): Promise<any> {
  const fileContent = await exports.readFile(source, 'utf8');

  try {
    return JSON.parse(fileContent);
  } catch (error) {
    throw new Error(`Invalid json file (${source})`);
  }
};
