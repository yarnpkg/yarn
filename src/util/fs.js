/* @flow */

import BlockingQueue from './blocking-queue.js';
import * as promise from './promise.js';
import {promisify} from './promise.js';
import map from './map.js';

const path = require('path');
const fs = require('fs');

export const lockQueue = new BlockingQueue('fs lock');

export const readFileBuffer = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);
export const readlink = promisify(fs.readlink);
export const realpath = promisify(fs.realpath);
export const readdir = promisify(fs.readdir);
export const rename = promisify(fs.rename);
export const access = promisify(fs.access);
export const stat = promisify(fs.stat);
export const unlink = promisify(require('rimraf'));
export const mkdirp = promisify(require('mkdirp'));
export const exists = promisify(fs.exists, true);
export const lstat = promisify(fs.lstat);
export const chmod = promisify(fs.chmod);

const fsSymlink = promisify(fs.symlink);
const invariant = require('invariant');
const stripBOM = require('strip-bom');

const noop = () => {};

export type CopyQueueItem = {
  src: string,
  dest: string,
  onFresh?: ?() => void,
  onDone?: ?() => void,
};

type CopyQueue = Array<CopyQueueItem>;

type CopyFileAction = {
  type: 'file',
  src: string,
  dest: string,
  atime: number,
  mtime: number,
  mode: number
};

type CopySymlinkAction = {
  type: 'symlink',
  dest: string,
  linkname: string,
};

type CopyActions = Array<CopyFileAction | CopySymlinkAction>;

type PossibleExtraneous = void | false | Iterable<string>;

type CopyOptions = {
  onProgress: (dest: string) => void,
  onStart: (num: number) => void,
  possibleExtraneous: PossibleExtraneous,
  ignoreBasenames: Array<string>,
};

async function buildActionsForCopy(
  queue: CopyQueue,
  events: CopyOptions,
  possibleExtraneousSeed: PossibleExtraneous,
): Promise<CopyActions> {
  const possibleExtraneous: Set<string> = new Set(possibleExtraneousSeed || []);
  const noExtraneous = possibleExtraneousSeed === false;
  const files: Set<string> = new Set();

  // initialise events
  for (const item of queue) {
    const onDone = item.onDone;
    item.onDone = () => {
      events.onProgress(item.dest);
      if (onDone) {
        onDone();
      }
    };
  }
  events.onStart(queue.length);

  // start building actions
  const actions: CopyActions = [];

  // custom concurrency logic as we're always executing stacks of 4 queue items
  // at a time due to the requirement to push items onto the queue
  while (queue.length) {
    const items = queue.splice(0, 4);
    await Promise.all(items.map(build));
  }

  // remove all extraneous files that weren't in the tree
  if (!noExtraneous) {
    for (const loc of possibleExtraneous) {
      if (!files.has(loc)) {
        await unlink(loc);
      }
    }
  }

  return actions;

  //
  async function build(data): Promise<void> {
    const {src, dest} = data;
    const onFresh = data.onFresh || noop;
    const onDone = data.onDone || noop;
    files.add(dest);

    if (events.ignoreBasenames.indexOf(path.basename(src)) >= 0) {
      // ignored file
      return;
    }

    const srcStat = await lstat(src);
    let srcFiles;

    if (srcStat.isDirectory()) {
      srcFiles = await readdir(src);
    }

    if (await exists(dest)) {
      const destStat = await lstat(dest);

      const bothSymlinks = srcStat.isSymbolicLink() && destStat.isSymbolicLink();
      const bothFolders = srcStat.isDirectory() && destStat.isDirectory();
      const bothFiles = srcStat.isFile() && destStat.isFile();

      if (srcStat.mode !== destStat.mode) {
        try {
          await access(dest, srcStat.mode);
        } catch (err) {
          // EINVAL access errors sometimes happen which shouldn't because node shouldn't be giving
          // us modes that aren't valid. investigate this, it's generally safe to proceed.
        }
      }

      if (bothFiles && srcStat.size === destStat.size && +srcStat.mtime === +destStat.mtime) {
        // we can safely assume this is the same file
        onDone();
        return;
      }

      if (bothSymlinks && await readlink(src) === await readlink(dest)) {
        // if both symlinks are the same then we can continue on
        onDone();
        return;
      }

      if (bothFolders && !noExtraneous) {
        // mark files that aren't in this folder as possibly extraneous
        const destFiles = await readdir(dest);
        invariant(srcFiles, 'src files not initialised');

        for (const file of destFiles) {
          if (srcFiles.indexOf(file) < 0) {
            const loc = path.join(dest, file);
            possibleExtraneous.add(loc);

            if ((await lstat(loc)).isDirectory()) {
              for (const file of await readdir(loc)) {
                possibleExtraneous.add(path.join(loc, file));
              }
            }
          }
        }
      }
    }

    if (srcStat.isSymbolicLink()) {
      onFresh();
      const linkname = await readlink(src);
      actions.push({
        type: 'symlink',
        dest,
        linkname,
      });
      onDone();
    } else if (srcStat.isDirectory()) {
      await mkdirp(dest);

      const destParts = dest.split(path.sep);
      while (destParts.length) {
        files.add(destParts.join(path.sep));
        destParts.pop();
      }

      // push all files to queue
      invariant(srcFiles, 'src files not initialised');
      let remaining = srcFiles.length;
      if (!remaining) {
        onDone();
      }
      for (const file of srcFiles) {
        queue.push({
          onFresh,
          src: path.join(src, file),
          dest: path.join(dest, file),
          onDone: () => {
            if (--remaining === 0) {
              onDone();
            }
          },
        });
      }
    } else if (srcStat.isFile()) {
      onFresh();
      actions.push({
        type: 'file',
        src,
        dest,
        atime: srcStat.atime,
        mtime: srcStat.mtime,
        mode: srcStat.mode,
      });
      onDone();
    } else {
      throw new Error(`unsure how to copy this: ${src}`);
    }
  }
}

export function copy(src: string, dest: string): Promise<void> {
  return copyBulk([{src, dest}]);
}

export async function copyBulk(
  queue: CopyQueue,
  _events?: {
    onProgress?: ?(dest: string) => void,
    onStart?: ?(num: number) => void,
    possibleExtraneous?: PossibleExtraneous,
    ignoreBasenames?: Array<string>,
  },
): Promise<void> {
  const events: CopyOptions = {
    onStart: (_events && _events.onStart) || noop,
    onProgress: (_events && _events.onProgress) || noop,
    possibleExtraneous: _events ? _events.possibleExtraneous : [],
    ignoreBasenames: (_events && _events.ignoreBasenames) || [],
  };

  const actions: CopyActions = await buildActionsForCopy(queue, events, events.possibleExtraneous);
  events.onStart(actions.length);

  const fileActions = actions.filter((action): boolean => action.type === 'file');
  await promise.queue(fileActions, (data): Promise<void> => new Promise((resolve, reject) => {
    const readStream = fs.createReadStream(data.src);
    const writeStream = fs.createWriteStream(data.dest, {mode: data.mode});

    readStream.on('error', reject);
    writeStream.on('error', reject);

    writeStream.on('open', function() {
      readStream.pipe(writeStream);
    });

    writeStream.once('finish', function() {
      fs.utimes(data.dest, data.atime, data.mtime, function(err) {
        if (err) {
          reject(err);
        } else {
          events.onProgress(data.dest);
          resolve();
        }
      });
    });
  }), 4);

  // we need to copy symlinks last as the could reference files we were copying
  const symlinkActions = actions.filter((action): boolean => action.type === 'symlink');
  await promise.queue(symlinkActions, (data): Promise<void> => {
    return symlink(path.resolve(path.dirname(data.dest), data.linkname), data.dest);
  });
}

function _readFile(loc: string, encoding: string): Promise<any> {
  return new Promise((resolve, reject) => {
    fs.readFile(loc, encoding, function(err, content) {
      if (err) {
        reject(err);
      } else {
        resolve(content);
      }
    });
  });
}

export function readFile(loc: string): Promise<string> {
  return _readFile(loc, 'utf8').then(normalizeOS);
}

export function readFileRaw(loc: string): Promise<Buffer> {
  return _readFile(loc, 'binary');
}

export async function readFileAny(files: Array<string>): Promise<?string> {
  for (const file of files) {
    if (await exists(file)) {
      return readFile(file);
    }
  }
  return null;
}

export async function readJson(loc: string): Promise<Object> {
  const file = await readFile(loc);
  try {
    return map(JSON.parse(stripBOM(file)));
  } catch (err) {
    err.message = `${loc}: ${err.message}`;
    throw err;
  }
}

export async function find(filename: string, dir: string): Promise<string | false> {
  const parts = dir.split(path.sep);

  while (parts.length) {
    const loc = parts.concat(filename).join(path.sep);

    if (await exists(loc)) {
      return loc;
    } else {
      parts.pop();
    }
  }

  return false;
}

export async function symlink(src: string, dest: string): Promise<void> {
  try {
    const stats = await lstat(dest);

    if (stats.isSymbolicLink() && await exists(dest)) {
      const resolved = await realpath(dest);
      if (resolved === src) {
        return;
      }
    }

    await unlink(dest);
  } catch (err) {
    if (err.code !== 'ENOENT') {
      throw err;
    }
  }

  try {
    if (process.platform === 'win32') {
      // use directory junctions if possible on win32, this requires absolute paths
      await fsSymlink(src, dest, 'junction');
    } else {
      // use relative paths otherwise which will be retained if the directory is moved
      const relative = path.relative(path.dirname(dest), src);
      await fsSymlink(relative, dest);
    }
  } catch (err) {
    if (err.code === 'EEXIST') {
      // race condition
      await symlink(src, dest);
    } else {
      throw err;
    }
  }
}

export type WalkFiles = Array<{
  relative: string,
  absolute: string,
  basename: string,
  mtime: number,
}>;

export async function walk(
  dir: string,
  relativeDir?: ?string,
  ignoreBasenames?: Set<string> = new Set(),
): Promise<WalkFiles> {
  let files = [];

  let filenames = await readdir(dir);
  if (ignoreBasenames.size) {
    filenames = filenames.filter((name) => !ignoreBasenames.has(name));
  }

  for (const name of filenames) {
    const relative = relativeDir ? path.join(relativeDir, name) : name;
    const loc = path.join(dir, name);
    const stat = await lstat(loc);

    files.push({
      relative,
      basename: name,
      absolute: loc,
      mtime: +stat.mtime,
    });

    if (stat.isDirectory()) {
      files = files.concat(await walk(loc, relative, ignoreBasenames));
    }
  }

  return files;
}

export function normalizeOS(body: string): string {
  return body.replace(/\r\n/g, '\n');
}
