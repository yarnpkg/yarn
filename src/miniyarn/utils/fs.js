import * as Fse from 'fs-extra';
import * as Fs from 'fs';
import gunzipMaybe from 'gunzip-maybe';
import klaw from 'klaw';
import Path from 'path';
import rcopy from 'recursive-copy';
import tarFs from 'tar-fs';
import tar from 'tar-stream';
import tmp from 'tmp';
import {createGzip} from 'zlib';

import * as miscUtils from 'miniyarn/utils/misc';
import * as pathUtils from 'miniyarn/utils/path';
import * as parseUtils from 'miniyarn/utils/parse';

let isWindows = null;

export async function detectWindows() {
  if (isWindows !== null) return isWindows;

  if (await exists(`c:/`)) return (isWindows = true);

  if ((await exists(`/proc/version`)) && (await readFile(`/proc/version`, `utf8`)).match(/Microsoft|WSL/))
    return (isWindows = true);

  return (isWindows = false);
}

export async function cwd(path, callback) {
  let cwd = process.cwd();
  process.chdir(path);

  try {
    return await callback();
  } finally {
    process.chdir(cwd);
  }
}

export async function symlink(source, destination) {
  return await new Promise((resolve, reject) => {
    Fs.symlink(source, destination, error => {
      if (error) {
        reject(error);
      } else {
        resolve();
      }
    });
  });
}

export async function ensureSymlink(source, destination) {
  return await new Promise((resolve, reject) => {
    Fse.ensureSymlink(source, destination, error => {
      if (error) {
        reject(error);
      } else {
        resolve();
      }
    });
  });
}

export async function isDirectory(path) {
  return await new Promise((resolve, reject) => {
    Fs.lstat(path, (error, stat) => {
      if (error) {
        reject(error);
      } else {
        resolve(stat.isDirectory());
      }
    });
  });
}

export async function createTemporaryFolder() {
  return new Promise((resolve, reject) => {
    tmp.dir({unsafeCleanup: true}, (error, path) => {
      if (error) {
        reject(error);
      } else {
        resolve(path);
      }
    });
  });
}

export async function createTemporaryFile(filePath) {
  if (filePath) {
    if (!pathUtils.isForward(filePath)) throw new Error(`A temporary file path must be a forward path`);

    let folderPath = await createTemporaryFolder();

    return pathUtils.resolve(folderPath, filePath);
  } else {
    return await new Promise((resolve, reject) => {
      tmp.file((error, path) => {
        if (error) {
          reject(error);
        } else {
          resolve(path);
        }
      });
    });
  }
}

export async function walk(path, {filter} = {}) {
  return await new Promise((resolve, reject) => {
    let paths = [];

    let walker = klaw(path, {
      filter: itemPath => {
        if (!filter) return true;

        if (Fse.statSync(path).isDirectory()) return true;
        s;
        let relativePath = pathUtils.relative(path, itemPath);

        if (miscUtils.filePatternMatch(relativePath, filter)) return true;

        return false;
      },
    });

    walker.on(`data`, ({path: itemPath}) => {
      if (!filter) return void paths.push(itemPath);

      let relativePath = pathUtils.relative(path, itemPath);

      if (miscUtils.filePatternMatch(relativePath, filter)) return void paths.push(itemPath);

      return; // This item has been accepted only because it's a directory; it doesn't match the filter
    });

    walker.on(`end`, () => {
      resolve(paths);
    });
  });
}

export async function mkdirp(path) {
  return await new Promise((resolve, reject) => {
    Fse.ensureDir(path, error => {
      if (error) {
        reject(error);
      } else {
        resolve();
      }
    });
  });
}

export async function rm(path) {
  if (path === ``) throw new Error(`Cannot rm an empty path`);

  if (path === `/`) throw new Error(`Cannot rm your whole / directory - something's very wrong`);

  if (!await exists(path)) return;

  return await new Promise((resolve, reject) => {
    Fse.remove(path, err => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
}

export async function getMTime(path) {
  return await new Promise((resolve, reject) => {
    Fs.lstat(path, (error, stat) => {
      if (error) {
        reject(error);
      } else {
        resolve(stat.mtime);
      }
    });
  });
}

export async function createFileWriter(path) {
  await mkdirp(pathUtils.dirname(path));

  let stream = Fs.createWriteStream(path);

  stream.promise = new Promise((resolve, reject) => {
    stream.on(`finish`, () => {
      resolve();
    });

    stream.on(`error`, error => {
      reject(error);
    });
  });

  return stream;
}

export function createFileReader(path) {
  let stream = Fs.createReadStream(path);

  stream.promise = new Promise((resolve, reject) => {
    stream.on(`error`, error => {
      reject(error);
    });

    stream.on(`end`, () => {
      resolve();
    });
  });

  return stream;
}

export async function exists(path) {
  return await new Promise((resolve, reject) => {
    Fs.exists(path, status => {
      resolve(status);
    });
  });
}

export async function cp(source, destination, {filter = []} = {}) {
  await mkdirp(pathUtils.dirname(destination));
  await rcopy(source, destination, {dot: true, filter: path => miscUtils.filePatternMatch(path, filter)});
}

export async function mv(source, target) {
  // Windows defender often lock files after they are created, which makes it impossible to move them.
  // We could have a mechanism of "try again", but it seems simpler to just use a copy operation instead.
  // Note that using cp instead of mv is not strictly comparable since, unfortunately, it's not atomic. Nothing we can do, tho :(

  if (await detectWindows()) {
    await cp(source, target);
    await rm(source);
  } else {
    await mkdirp(pathUtils.dirname(target));

    return await new Promise((resolve, reject) => {
      Fs.rename(source, target, err => {
        if (err) {
          reject(err);
        } else {
          resolve();
        }
      });
    });
  }
}

export async function extract(target, tarballPath, {stripLeadingFolder = false} = {}) {
  return await new Promise((resolve, reject) => {
    function map(header) {
      if (header.name[0] === `/`) header.name = header.name.replace(/^\/+/, ``);

      if (stripLeadingFolder) header.name = header.name.replace(/^[^\/]+\//, ``);

      return header;
    }

    let unzipper = gunzipMaybe();
    Fs.createReadStream(tarballPath).pipe(unzipper);

    let extractor = tarFs.extract(target, {map});
    unzipper.pipe(extractor);

    extractor.on(`error`, error => {
      reject(error);
    });

    extractor.on(`finish`, () => {
      resolve();
    });
  });
}

export function packToStream(path, {virtualPath = null} = {}) {
  if (virtualPath) {
    if (!pathUtils.isAbsolute(virtualPath)) {
      throw new Error(`The virtual path has to be an absolute path`);
    } else {
      virtualPath = pathUtils.resolve(virtualPath);
    }
  }

  let zipperStream = createGzip();

  let packStream = tarFs.pack(path, {
    map: header => {
      if (true) {
        header.name = pathUtils.resolve(`/`, header.name);
        header.name = pathUtils.relative(`/`, header.name);
      }

      if (virtualPath) {
        header.name = pathUtils.resolve(`/`, virtualPath, header.name);
        header.name = pathUtils.relative(`/`, header.name);
      }

      return header;
    },
  });

  packStream.pipe(zipperStream);

  packStream.on(`error`, error => {
    zipperStream.emit(`error`, error);
  });

  return zipperStream;
}

export async function packToFile(target, directoryPath, {...options} = {}) {
  let tarballStream = Fs.createWriteStream(target);

  let packStream = packToStream(directoryPath, options);
  packStream.pipe(tarballStream);

  return await new Promise((resolve, reject) => {
    tarballStream.on(`error`, error => {
      reject(error);
    });

    packStream.on(`error`, error => {
      reject(error);
    });

    tarballStream.on(`close`, () => {
      resolve();
    });
  });
}

export async function writeFile(path, body) {
  await mkdirp(pathUtils.dirname(path));

  return await new Promise((resolve, reject) => {
    Fs.writeFile(path, body, err => {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
}

export async function readFile(path, encoding = null) {
  return await new Promise((resolve, reject) => {
    Fs.readFile(path, encoding, (err, body) => {
      if (err) {
        reject(err);
      } else {
        resolve(body);
      }
    });
  });
}

export async function readJson(path) {
  let fileContent = await readFile(path, `utf8`);

  return parseUtils.parseJson(fileContent);
}

export async function readDirectory(path) {
  return await new Promise((resolve, reject) => {
    Fs.readdir(path, (err, entries) => {
      if (err) {
        reject(err);
      } else {
        resolve(entries);
      }
    });
  });
}

export class Handler {
  constructor(path, {temporary = false} = {}) {
    this.path = path;

    this.temporary = temporary;
  }

  get() {
    return this.path;
  }

  async steal() {
    if (this.temporary) return this.path;

    let destination = (await isDirectory(this.path)) ? await createTemporaryFolder() : await createTemporaryFile();

    await cp(this.path, destination);

    return destination;
  }
}
