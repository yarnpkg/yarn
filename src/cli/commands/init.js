/* @flow */

import type {Reporter} from '../../reporters/index.js';
import type Config from '../../config.js';
import {stringifyPerson} from '../../util/normalize-manifest/util.js';
import {registryNames} from '../../registries/index.js';
import Lockfile from '../../lockfile/wrapper.js';
import {Install} from './install.js';
import * as child from '../../util/child.js';
import * as fs from '../../util/fs.js';

const objectPath = require('object-path');
const path = require('path');

export function setFlags(commander: Object) {
  commander.option('-y, --yes', 'use default options');
}

export async function run(
  config: Config,
  reporter: Reporter,
  flags: Object,
  args: Array<string>,
): Promise<void> {
  const lockfile = new Lockfile();
  const install = new Install(flags, config, reporter, lockfile);
  const manifests = await install.getRootManifests();

  let gitUrl;
  const author = {
    name: config.getOption('init-author-name'),
    email: config.getOption('init-author-email'),
    url: config.getOption('init-author-url'),
  };
  if (await fs.exists(path.join(config.cwd, '.git'))) {
    // get git origin of the cwd
    try {
      gitUrl = await child.spawn('git', ['config', 'remote.origin.url'], {cwd: config.cwd});
    } catch (ex) {
      // Ignore - Git repo may not have an origin URL yet (eg. if it only exists locally)
    }

    // get author default based on git config
    author.name = author.name || await child.spawn('git', ['config', 'user.name']);
    author.email = author.email || await child.spawn('git', ['config', 'user.email']);
  }

  const keys = [
    {
      key: 'name',
      question: 'name',
      default: path.basename(config.cwd),
    },
    {
      key: 'version',
      question: 'version',
      default: String(config.getOption('init-version')),
    },
    {
      key: 'description',
      question: 'description',
      default: '',
    },
    {
      key: 'main',
      question: 'entry point',
      default: 'index.js',
    },
    {
      key: 'repository.url',
      question: 'git repository',
      default: gitUrl,
    },
    {
      key: 'author',
      question: 'author',
      default: stringifyPerson(author),
    },
    {
      key: 'license',
      question: 'license',
      default: String(config.getOption('init-license')),
    },
  ];

  // get answers
  const pkg = {};
  for (const entry of keys) {
    const {yes} = flags;
    const {key: manifestKey} = entry;
    let {question, default: def} = entry;

    for (const registryName of registryNames) {
      const {object} = manifests[registryName];
      const val = objectPath.get(object, manifestKey);
      if (val) {
        def = val;
        break;
      }
    }

    if (def) {
      question += ` (${def})`;
    }

    let answer;

    if (yes) {
      answer = def;
    } else {
      answer = (await reporter.question(question)) || def;
    }

    if (answer) {
      objectPath.set(pkg, manifestKey, answer);
    }
  }

  // if we have a git url then set the type
  if (pkg.repository && pkg.repository.url) {
    pkg.repository.type = 'git';
  }

  // save answers
  const targetManifests = [];
  for (const registryName of registryNames) {
    const info = manifests[registryName];
    if (info.exists) {
      targetManifests.push(info);
    }
  }
  if (!targetManifests.length) {
    targetManifests.push(manifests.npm);
  }
  for (const targetManifest of targetManifests) {
    Object.assign(targetManifest.object, pkg);
    reporter.success(`Saved ${path.basename(targetManifest.loc)}`);
  }

  await install.saveRootManifests(manifests);
}
