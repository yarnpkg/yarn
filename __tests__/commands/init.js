/* @flow */

import type {QuestionOptions} from '../../src/reporters/types.js';
import {ConsoleReporter} from '../../src/reporters/index.js';
import {run as buildRun} from './_helpers.js';
import {getGitConfigInfo, run as runInit} from '../../src/cli/commands/init.js';
import * as fs from '../../src/util/fs.js';

const path = require('path');

jasmine.DEFAULT_TIMEOUT_INTERVAL = 60000;

const fixturesLoc = path.join(__dirname, '..', 'fixtures', 'init');

test.concurrent('init --yes should create package.json with defaults', (): Promise<void> => {
  return buildRun(
    ConsoleReporter,
    fixturesLoc,
    (args, flags, config, reporter, lockfile): Promise<void> => {
      return runInit(config, reporter, flags, args);
    },
    [],
    {yes: true},
    'init-yes',
    async (config): Promise<void> => {
      const {cwd} = config;
      const manifestFile = await fs.readFile(path.join(cwd, 'package.json'));
      const manifest = JSON.parse(manifestFile);

      // Name is derived from directory name which is dynamic so check
      // that separately and then remove from snapshot
      expect(manifest.name).toEqual(path.basename(cwd));
      expect(manifest.private).toEqual(undefined);
      expect({...manifest, name: 'init-yes'}).toMatchSnapshot('init-yes');
    },
  );
});

test.concurrent('init --yes --private should create package.json with defaults and private true', (): Promise<void> => {
  return buildRun(
    ConsoleReporter,
    fixturesLoc,
    (args, flags, config, reporter, lockfile): Promise<void> => {
      return runInit(config, reporter, flags, args);
    },
    [],
    {yes: true, private: true},
    'init-yes-private',
    async (config): Promise<void> => {
      const {cwd} = config;
      const manifestFile = await fs.readFile(path.join(cwd, 'package.json'));
      const manifest = JSON.parse(manifestFile);

      // Name is derived from directory name which is dynamic so check
      // that separately and then remove from snapshot
      expect(manifest.name).toEqual(path.basename(cwd));
      expect(manifest.private).toEqual(true);
      expect({...manifest, name: 'init-yes-private'}).toMatchSnapshot('init-yes-private');
    },
  );
});

test.concurrent('init using Github shorthand should resolve to full repository URL', (): Promise<void> => {
  const questionMap = Object.freeze({
    name: 'hi-github',
    version: '',
    description: '',
    'entry point': '',
    'repository url': 'user/repo',
    author: '',
    license: '',
    private: 'false',
  });
  class TestReporter extends ConsoleReporter {
    question(question: string, options?: QuestionOptions = {}): Promise<string> {
      return new Promise((resolve, reject) => {
        const parsedQuestion = question.replace(/ \((.*?)\)/g, '');
        if (parsedQuestion in questionMap) {
          resolve(questionMap[parsedQuestion]);
        } else {
          reject(new Error(`Question not found in question-answer map ${parsedQuestion}`));
        }
      });
    }
  }

  return buildRun(
    TestReporter,
    fixturesLoc,
    (args, flags, config, reporter, lockfile): Promise<void> => {
      return runInit(config, reporter, flags, args);
    },
    [],
    {},
    'init-github',
    async (config): Promise<void> => {
      const manifestFile = await fs.readFile(path.join(config.cwd, 'package.json'));

      expect(JSON.parse(manifestFile)).toMatchSnapshot('init-github');
    },
  );
});

test.concurrent('init and give private empty', (): Promise<void> => {
  const questionMap = Object.freeze({
    name: 'private-empty',
    version: '',
    description: '',
    'entry point': '',
    'repository url': '',
    author: '',
    license: '',
    private: '',
  });
  class TestReporter extends ConsoleReporter {
    question(question: string, options?: QuestionOptions = {}): Promise<string> {
      return new Promise((resolve, reject) => {
        const parsedQuestion = question.replace(/ \((.*?)\)/g, '');
        if (parsedQuestion in questionMap) {
          resolve(questionMap[parsedQuestion]);
        } else {
          reject(new Error(`Question not found in question-answer map ${parsedQuestion}`));
        }
      });
    }
  }

  return buildRun(
    TestReporter,
    fixturesLoc,
    (args, flags, config, reporter, lockfile): Promise<void> => {
      return runInit(config, reporter, flags, args);
    },
    [],
    {},
    'init-github',
    async (config): Promise<void> => {
      const manifestFile = await fs.readFile(path.join(config.cwd, 'package.json'));
      expect(JSON.parse(manifestFile)).toMatchSnapshot('init-private-empty');
    },
  );
});

test.concurrent('getGitConfigInfo should not return the git config val', async (): Promise<void> => {
  expect('hi seb').toEqual(await getGitConfigInfo('some-info', () => Promise.resolve('hi seb')));
});

test.concurrent('getGitConfigInfo should not fail when git fails', async (): Promise<void> => {
  expect('').toEqual(await getGitConfigInfo('some-info', () => Promise.reject(Error())));
});
