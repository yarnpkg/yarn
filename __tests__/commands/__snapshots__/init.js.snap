// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`init-github 1`] = `
Object {
  "license": "MIT",
  "main": "index.js",
  "name": "hi-github",
  "private": false,
  "repository": "https://github.com/user/repo",
  "version": "1.0.0",
}
`;

exports[`init-private-empty 1`] = `
Object {
  "license": "MIT",
  "main": "index.js",
  "name": "private-empty",
  "version": "1.0.0",
}
`;

exports[`init-yes 1`] = `
Object {
  "license": "MIT",
  "main": "index.js",
  "name": "init-yes",
  "version": "1.0.0",
}
`;

exports[`init-yes-private 1`] = `
Object {
  "license": "MIT",
  "main": "index.js",
  "name": "init-yes-private",
  "private": true,
  "version": "1.0.0",
}
`;
